src/20240205193700-merge-grouped-exercise-into-exercise-group.ts

```diff
@@ -0,0 +1,615 @@
+
+const ChildContentDecoder = t.type({
+  plugin: t.literal('exercise'),
+  state: t.intersection([
+    t.type({
+      content: t.type({
+        plugin: t.literal('rows'),
+      }),
+    }),
+    t.partial({ licenseId: t.number }),
+  ]),
+})
+const ParentContentDecoder = t.type({
+  plugin: t.literal('type-text-exercise-group'),
+  state: t.record(t.string, t.unknown),
+})
+const RowPluginDecoder = t.type({ plugin: t.literal('rows') })
+const TextPluginDecoder = t.type({ plugin: t.literal('text') })
+
+createMigration(exports, {
+  up: async (db) => {
+    const apiCache = new ApiCache({ enableLogging: false })
+
+    const batchSize = 5000
+    let entities: Row[] = []
+
+    do {
+      const lastUpdatedEntityId = entities.at(-1)?.entityId ?? 0
+
+      entities = await db.runSql<Row[]>(
+        `select
+           entity.id as entityId
+         from entity
+         join type on entity.type_id = type.id
+         where type.name in ("text-exercise-group")
+         and entity.id > ?
+         order by entity.id limit ?`,
+        lastUpdatedEntityId,
+        batchSize,
+      )
+
+      for (const entity of entities) {
+        await updateExerciseGroup(db, apiCache, entity.entityId)
+      }
+    } while (entities.length > 0)
+
+    await apiCache.deleteUnrevisedRevisions()
+    await apiCache.quit()
+
+    interface Row {
+      entityId: number
+    }
+  },
+})
+
+async function updateExerciseGroup(
+  db: Database,
+  apiCache: ApiCache,
+  parentId: number,
+) {
+  // Load the tree of the parent entity. It is either only the parent (without the
+  // solution / without any children):
+  //
+  //    [ parent ]
+  //
+  // or it is a parent with children:
+  //
+  //    [ parent ]
+  //      └ [ child1 ]
+  //      └ [ child2 ]
+  //
+  // Next to the entity in the node the list of revisions are stored:
+  //
+  //    [ parent ]:      [ e1, e2, e3, ...]
+  //      └ [ child1 ]   [ s1, s2, s3, ...]
+  const parentNode = await loadEntityNode(db, parentId)
+
+  for (const child of parentNode.children) {
+    assert(child.children.length === 0, 'child must not have children')
+  }
+
+  // When a parent does not have any revision there is no transformation
+  // possible even if the solution has revisions.
+  if (parentNode.value.revisions.length === 0) return
+
+  // From the tree of revisions we calculate the list of overall edits. We
+  // represent an edit as a tree with the
+  // same structure as the parent. The value for parent / child is
+  // a revision (type `EntityWithRevision`) when the parent / child was
+  // edited or `null` when it was not edited. So let's assume the entity
+  // tree with the revisions has the structure
+  //
+  //    [ parent ]       [ edit_on_date1 | edit_on_date2 ]
+  //      └ [ child1 ]   [ edit_on_date1 | edit_on_date3 ]
+  //
+  // Then the list of edits are
+  //
+  // [ edit_on_date1 ]     | [ edit_on_date2 ] | [ null ]
+  //   └ [ edit_on_date2 ] |   └ [ null ]      |   └ [ edit_on_date3 ]
```

Comment:
```suggestion
  //   └ [ edit_on_date1 ] |   └ [ null ]      |   └ [ edit_on_date3 ]
```
Do you mean this? I may not have understood it.